include/pybind11/warnings.h

```diff
@@ -0,0 +1,70 @@
+/*
+    pybind11/warnings.h: Python warnings wrappers.
+
+    Copyright (c) 2024 Jan Iwaszkiewicz <jiwaszkiewicz6@gmail.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "pybind11.h"
+#include "detail/common.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+inline bool PyWarning_Check(PyObject *obj) {
+    int result = PyObject_IsSubclass(obj, PyExc_Warning);
+    if (result == 1) {
+        return true;
+    }
+    if (result == -1) {
+        raise_from(PyExc_SystemError,
+                   "PyWarning_Check(): internal error of Python C API while "
+                   "checking a subclass of the object!");
+        throw error_already_set();
+    }
+    return false;
+}
+
+PYBIND11_NAMESPACE_END(detail)
+
+PYBIND11_NAMESPACE_BEGIN(warnings)
+
+inline object
+new_warning_type(handle scope, const char *name, handle base = PyExc_RuntimeWarning) {
+    if (!detail::PyWarning_Check(base.ptr())) {
+        pybind11_fail("warning(): cannot create custom warning, base must be a subclass of "
+                      "PyExc_Warning!");
+    }
+    if (hasattr(scope, "__dict__") && scope.attr("__dict__").contains(name)) {
+        pybind11_fail("Error during initialization: multiple incompatible "
+                      "definitions with name \""
+                      + std::string(name) + "\"");
+    }
+    std::string full_name = scope.attr("__name__").cast<std::string>() + std::string(".") + name;
+    handle h(PyErr_NewException(const_cast<char *>(full_name.c_str()), base.ptr(), nullptr));
```

Comment:
I believe the `PyErr_NewException()` documentation is incomplete because there is no mention of errors, but almost certainly we need at least this:
```
if (!h) {
    throw error_already_set();
}
```
I think that'd be sufficient, although `raise_from()` like you have above might be even better. I'm not sure how to judge cost:benefit here.

Reply:
I agree on that, I have modified it and move creation of handle after the check -- so any null should be verified first.

About the docs... yes, they are "very vague" about error messages so referring to [them](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_NewException) and [cpython code](https://github.com/python/cpython/blob/1054a755a3016f95fcd24b3ad20e8ed9048b7939/Python/errors.c#L1231) this is the best that can be done. However, I am not able to think of a way of testing it. Seems like there is no way to break it. To break it down:
* `name` is always consisting one dot - it's forced by `full_name`.
* `base` is also validated by `PyWarning_Check` (plus it could also be null by design of cpython function).
* `dict` by itself is always a nullptr here so that check in cpython is automatically "passed".
I can only imagine if somehow `modulename` is broken. Maybe I am missing something but it seems to me like trying to break just to prove "that it's breakable". Do you have any thoughts on that?

Reply:
It's very common that testing of error conditions is incomplete. It's often just too difficult.

What I usually do: temporarily change e.g. `if (ptr == nullptr)` to `if (ptr != nullptr)` so that the `if` branch is reached, run the tests to convince myself that it works as intended, then change it back.